File: cpp/src/system/variables/wavespeeds.cpp
#include "../objects/gpr_objects.h"
#include "derivatives.h"
#include "eos.h"
#include <cmath>

double c_0(double ρ, double p, Mat3_3r A, Par &MP) {
  // Returns the adiabatic sound speed for the MG EOS
  double dE_dρ = dEdρ(ρ, p, A, MP);
  double dE_dp = dEdp(ρ, MP);
  return sqrt((p / (ρ * ρ) - dE_dρ) / dE_dp);
}

double c_h(double ρ, double T, Par &MP) {
  // Returns the velocity of the heat characteristic at equilibrium
  double cα2 = MP.cα2;
  double cv = MP.cv;
  return sqrt(cα2 * T / cv) / ρ;
}

File: python/models/gpr/variables/wavespeeds.py
from numpy import sqrt

from models.gpr.variables.derivatives import dEdρ, dEdp


def c_0(ρ, p, A, MP):
    """ Returns the adiabatic sound speed for the Mie-Gruneisen EOS
    """
    dE_dρ = dEdρ(ρ, p, A, MP)
    dE_dp = dEdp(ρ, MP)
    return sqrt((p / ρ**2 - dE_dρ) / dE_dp)


def c_inf(ρ, p, MP):
    """ Returns the longitudinal characteristic speed
    """
    c0 = c_0(ρ, p, MP)
    cs2 = c_s2(ρ, MP)
    return sqrt(c0**2 + 4 / 3 * cs2)


def c_h(ρ, T, MP):
    """ Returns the velocity of the heat characteristic at equilibrium
    """
    cα2 = MP.cα2
    cv = MP.cv
    return sqrt(cα2 * T / cv) / ρ
